### PR TITLE
Make excpetions optional depending on build-time feature checks.

### PR DESCRIPTION
**Pull request by chuim**
Replaces all try/catch/throw keyword usage with a set of macros that are
appropriately defined in config.hpp based on the status of the
`cxx_exceptions` compiler feature.

Reported in issue #160.

### CODE REVIEW

File: immer/config.hpp [hunk 1]
```diff
@@ -18,6 +18,28 @@
 #endif
 #endif
 
+#ifdef __has_feature
+#if !__has_feature(cxx_exceptions)
```
Comment (chuim) [inline immer/config.hpp:22]: I tried having these two checks in a single line to directly choose the version of the macros to use, but that caused build errors like this:
```
/usr/local/google/home/carlosk/lib/immer/immer/config.hpp:21:45: error: missing binary operator before token "("
   21 | #if !defined(__has_feature) || __has_feature(cxx_exceptions)
      |    
```
Splitting into two separate lines like this worked.

**Review by arximboldi — COMMENTED** [review]
Overall LGTM. Have you run `clang-format` on it?

File: immer/detail/arrays/no_capacity.hpp [hunk 2]
```diff
@@ -159,10 +159,13 @@ struct no_capacity
     no_capacity push_back(T value) const
     {
         auto p = node_t::copy_n(size + 1, ptr, size);
-        IMMER_TRY {
+        IMMER_TRY
+        {
```
Comment (arximboldi) [inline immer/detail/arrays/no_capacity.hpp:163]: The new line before the brace looks weird. Do you think there is any way to convince `clang-format` to move the brace back into the previous line?

**Comment by arximboldi** [comment]
Cool, thanks! I think the last incarnation is acceptable. Merging!

**Pull request merged**